locustempus/main/views.py

```diff
@@ -118,22 +139,50 @@ def form_valid(self, form) -> HttpResponse:
         return result
 
 
-class CourseDetailView(LoggedInCourseMixin, DetailView):
-    model = Course
+class CourseDetailView(LoggedInCourseMixin, View):
     template_name = 'main/course_detail.html'
+    http_method_names = ['get', 'post']
 
-    def get_context_data(self, **kwargs):
-        ctx = super().get_context_data(**kwargs)
-        course: Course = kwargs.get('object')
-        ctx['projects'] = Project.objects.filter(course=course)\
-            .order_by('title')
-        ctx['is_faculty'] = course.is_true_faculty(self.request.user)
-        ctx['page_type'] = 'course'
-        ctx['breadcrumb'] = {
-            'Workspaces': reverse('course-list-view'),
-            course.title: '',
+    def post(self, request, *args, **kwargs):
+        course = get_object_or_404(Course, pk=kwargs.get('pk'))
+
+        is_grid = not request.session.get('project_list_grid', False)
+        request.session['project_list_grid'] = is_grid
+
+        ctx = {
+            'course': course,
+            'projects': Project.objects.filter(course=course)
+            .order_by('title'),
+            'is_faculty': course.is_true_faculty(self.request.user),
+            'page_type': 'course',
+            'project_list_grid': is_grid,
+            'breadcrumb': {
+                'Workspaces': reverse('course-list-view'),
+                course.title: '',
+            }
         }
-        return ctx
+        return render(request, self.template_name, ctx)
+
+    def get(self, request, *args, **kwargs):
+        course = get_object_or_404(Course, pk=kwargs.get('pk'))
+
+        is_grid = request.session.get('project_list_grid', True)
```

Comment:
The fetching of the grid state and setting it into the session is repeated a few times. Might be a good opportunity for a Mixin?

Reply:
I'm not sure. The session key is specific to the view, and the default value changes based on if it's a POST or a GET. I could pull it out, but it would just wrap `request.session.get`. Is there a different way to introduce the mixin?

Reply:
Yeah - I see your point, and am noting the differences now as I look back on it. Likely too much work / too little gain for an abstraction.